### PR TITLE
Add GetDefaultPkeyFromPci and GetPKeyByIndexFromPci functions

### PR DESCRIPTION
**Pull request by almaslennikov**
The functions read the from the PCI device's
infiniband dir and returns either the PKey for index0 or given index

### CODE REVIEW

**Review by ykulazhenkov — APPROVED** [review]

File: sriovnet.go [hunk 1]
```diff
@@ -520,3 +520,32 @@ func GetPciFromNetDevice(name string) (string, error) {
 	}
 	return base, nil
 }
+
+// GetPKeyFromPci returns the index0 PKey for the IB PCI device
+func GetPKeyFromPci(pciAddress string) (string, error) {
```
Comment (adrianchiris) [inline sriovnet.go:525]: this works for both PF and VFs ?

**Review by adrianchiris — APPROVED** [review]
LGTM assuming the answer to my comment is "yes"

later on we might want to split infiniband specific functionality to its own file if deemed nessecary.

File: sriovnet.go (hunk 1 above)
Comment (adrianchiris) [reply sriovnet.go:525]: there are several indexes for pkey membership, index0 is the default membership.

should we have two functions ?

`func GetPkeyByIndexFromPci(pciAddress string) (string, error)`

and

`func GetDefaultlPkeyFromPci(pciAddress string) (string, error`

Comment (almaslennikov) [reply sriovnet.go:525]: This works for both PFs and VFs as the mechanism around how PKeys are stored is the same for both.
I'm not opposed having two functions to retrieve different PKeys. Let me address that in the PR

Comment (almaslennikov) [reply sriovnet.go]: Split into two new functions

**Review by adrianchiris — APPROVED** [review]

**Comment by almaslennikov** [comment]
@SchSeba @moshe010 Could you take a look at this PR?

File: sriovnet.go [hunk 2]
```diff
@@ -520,3 +520,37 @@ func GetPciFromNetDevice(name string) (string, error) {
 	}
 	return base, nil
 }
+
+// GetPKeyByIndexFromPci returns the PKey stored under given index for the IB PCI device
+func GetPKeyByIndexFromPci(pciAddress string, index int) (string, error) {
+	pciDir := filepath.Join(PciSysDir, pciAddress, "infiniband")
+	dirEntries, err := utilfs.Fs.ReadDir(pciDir)
+	if err != nil {
+		return "", err
```
Comment (bn222) [inline sriovnet.go:529]: Le'ts be more specific
`
        return "", fmt.Errorf("failed to read infiniband directory: %v", err)
`

File: sriovnet.go [hunk 3]
```diff
@@ -520,3 +520,37 @@ func GetPciFromNetDevice(name string) (string, error) {
...
+	}
+	if len(dirEntries) == 0 {
+		return "", fmt.Errorf("infiniband ports directory is empty for device: %s", pciAddress)
+	}
+
+	index0FilePath := filepath.Join(portsDir, dirEntries[0].Name(), "pkeys", strconv.Itoa(index))
+	pKeyBytes, err := utilfs.Fs.ReadFile(index0FilePath)
+	if err != nil {
+		return "", err
```
Comment (bn222) [inline sriovnet.go:547]: Here as well:

`        return "", fmt.Errorf("failed to read PKey file: %v", err)
`

**Review by bn222 — COMMENTED** [review]
Two small nits that will aid debugging. Once those are addressed, it's an LGTM from me.

**Comment by almaslennikov** [comment]
Hi @bn222, addressed your comments.
@adrianchiris I think we can merge now

File: sriovnet.go [hunk 4]
```diff
@@ -520,3 +520,37 @@ func GetPciFromNetDevice(name string) (string, error) {
...
+		return "", fmt.Errorf("failed to read infiniband directory: %v", err)
+	}
+	if len(dirEntries) == 0 {
+		return "", fmt.Errorf("infiniband directory is empty for device: %s", pciAddress)
+	}
+
+	portsDir := filepath.Join(pciDir, dirEntries[0].Name(), "ports")
+	dirEntries, err = utilfs.Fs.ReadDir(portsDir)
+	if err != nil {
+		return "", fmt.Errorf("failed to infiniband ports directory: %v", err)
```
Comment (adrianchiris) [inline sriovnet.go:538]: nit: failed to read infiniband ports directory ?

File: sriovnet.go [hunk 5]
```diff
@@ -520,3 +520,37 @@ func GetPciFromNetDevice(name string) (string, error) {
...
+	}
+	if len(dirEntries) == 0 {
+		return "", fmt.Errorf("infiniband ports directory is empty for device: %s", pciAddress)
+	}
+
+	index0FilePath := filepath.Join(portsDir, dirEntries[0].Name(), "pkeys", strconv.Itoa(index))
```
Comment (adrianchiris) [inline sriovnet.go:544]: index0FilePath -> indexFilePath

File: sriovnet.go [hunk 6]
```diff
@@ -520,3 +520,37 @@ func GetPciFromNetDevice(name string) (string, error) {
 	}
 	return base, nil
 }
+
+// GetPKeyByIndexFromPci returns the PKey stored under given index for the IB PCI device
+func GetPKeyByIndexFromPci(pciAddress string, index int) (string, error) {
+	pciDir := filepath.Join(PciSysDir, pciAddress, "infiniband")
```
Comment (adrianchiris) [inline sriovnet.go:526]: question:

should we just try to read: `/sys/bus/pci/devices/<pciAddr>/infiniband/ports/1/pkeys/<index>` file straight away ?
just to save a bunch of filesystem operations.

also im not aware of devices that have multiple infiniband ports under the same PCI function.
(there was in the distant past but no more...)

Comment (almaslennikov) [reply sriovnet.go:526]: We actually need the name of the RDMA device to construct the path. However, I've updated the func to always use the port number 1

**Review by adrianchiris — APPROVED** [review]
LGTM,
@bn222 PTAL i believe we can merge this one

**Comment by bn222** [comment]
LGTM.

**Pull request merged**